network/p2p/tracer/gossipSubScoreTracer_test.go

```diff
@@ -0,0 +1,301 @@
+	scoreMetrics.On("OnMeshMessageDeliveredUpdated", topic1, mock.Anything).Return()
+	scoreMetrics.On("OnInvalidMessageDeliveredUpdated", topic1, mock.Anything).Return()
+	scoreMetrics.On("SetWarningStateCount", uint(0)).Return()
+
+	// 6. Subscribes the nodes to a common topic.
+	_, err := tracerNode.Subscribe(
+		topic1,
+		validator.TopicValidator(
+			unittest.Logger(),
+			unittest.AllowAllPeerFilter()))
+	require.NoError(t, err)
+
+	_, err = consensusNode.Subscribe(
+		topic1,
+		validator.TopicValidator(
+			unittest.Logger(),
+			unittest.AllowAllPeerFilter()))
+	require.NoError(t, err)
+
+	_, err = accessNode.Subscribe(
+		topic1,
+		validator.TopicValidator(
+			unittest.Logger(),
+			unittest.AllowAllPeerFilter()))
+	require.NoError(t, err)
+
+	// 7. Expects the tracer node to have the correct app scores, a non-zero score, an existing behaviour score, an existing
+	// IP score, and an existing mesh score.
+	assert.Eventually(t, func() bool {
+		// we expect the tracerNode to have the consensusNodes and accessNodes with the correct app scores.
+		exposer, ok := tracerNode.PeerScoreExposer()
+		require.True(t, ok)
+
+		score, ok := exposer.GetAppScore(consensusNode.Host().ID())
+		if !ok || score != consensusScore {
+			return false
+		}
+
+		score, ok = exposer.GetAppScore(accessNode.Host().ID())
+		if !ok || score != accessScore {
+			return false
+		}
+
+		// we expect the tracerNode to have the consensusNodes and accessNodes with a non-zero score.
+		score, ok = exposer.GetScore(consensusNode.Host().ID())
+		if !ok || score == 0 {
+			return false
+		}
+
+		score, ok = exposer.GetScore(accessNode.Host().ID())
+		if !ok || score == 0 {
+			return false
+		}
+
+		// we expect the tracerNode to have the consensusNodes and accessNodes with an existing behaviour score and ip score.
+		_, ok = exposer.GetBehaviourPenalty(consensusNode.Host().ID())
+		if !ok {
+			return false
+		}
+
+		_, ok = exposer.GetIPColocationFactor(consensusNode.Host().ID())
+		if !ok {
+			return false
+		}
+
+		_, ok = exposer.GetBehaviourPenalty(accessNode.Host().ID())
+		if !ok {
+			return false
+		}
+
+		_, ok = exposer.GetIPColocationFactor(accessNode.Host().ID())
+		if !ok {
+			return false
+		}
+
+		// we expect the tracerNode to have the consensusNodes and accessNodes with an existing mesh score.
+		consensusMeshScores, ok := exposer.GetTopicScores(consensusNode.Host().ID())
+		if !ok {
+			return false
+		}
+		_, ok = consensusMeshScores[topic1.String()]
+		if !ok {
+			return false
+		}
+
+		accessMeshScore, ok := exposer.GetTopicScores(accessNode.Host().ID())
+		if !ok {
+			return false
+		}
+		_, ok = accessMeshScore[topic1.String()]
+		return ok
+	}, 2*time.Second, 10*time.Millisecond)
+
+	time.Sleep(2 * time.Second)
+
+	// 8. Expects the score tracer to log the scores at least once.
+	assert.Eventually(t, func() bool {
+		return loggerCycle.Load() > 0
+	}, 2*time.Second, 10*time.Millisecond)
+}
+
+type mockPeerScoreMetrics struct {
```

Comment:
Why do we need this struct to wrap `mockmodule.NewGossipSubScoringMetrics(t)` instead of using it directly ?

Reply:
In the test for initializing a [libp2p node](https://github.com/onflow/flow-go/blob/1b5dcaf4c1a06fd77d3c2d8eb4f50be1354b6ef9/network/p2p/tracer/gossipSubScoreTracer_test.go#L79), we need to use the `module.GossipSubScoringMetrics` metric. However, we cannot use this metric alone to initialize the node because the node requires the parent interface of `module.NetworkMetrics`. Creating a mocked version of `module.NetworkMetrics` is difficult because we would have to mock all of its invocations throughout the node's components.

To simplify the test, we only want to mock the `module.GossipSubScoringMetrics` invocations and ignore the rest. To achieve this, we created `mockPeerScoreMetrics`, which wraps `module.GossipSubScoringMetrics` with `module.NoopCollector`, a no-op implementation. This allows us to mock only the gossipsub score collections without affecting the rest of the metrics collection and without breaking the test.